Add optional separate scheduler deployment

Timost opened the pull request:
This can be usefull when one wants to set different affinities for the
scheduler and the worker pods.

My personal use case is to take advantage of premptible nodes for dask workers but keep the scheduler on a guaranteed node.

Attempts to fix #208.

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -393,6 +396,73 @@ def __init__(
+        for pod_template, pod_type in [
+            (self.pod_template, "worker"),
+            (self.scheduler_pod_template, "scheduler"),
+        ]:
```
I wonder if you could call this function twice and pass these in as args. Would make it more readable.

Timost replied:
Yeah I see what you mean ! I'll make the modifications.

Timost replied:
The thing though is that there is a bit of logic to assign the scheduler template to the pod template by default. And this requires having knowledge of both templates. I can move this part out of the function, but I fell it belongs more there than in the main `_start` method.

jacobtomlinson replied:
I think my preference would be to have it in the main `_start` function. It makes the start function clearer to read.

Timost replied:
Ok I see what you mean. I'll change it.

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -393,6 +396,73 @@ def __init__(
+
+        self.scheduler_pod_template = clean_pod_template(
+            self.scheduler_pod_template, pod_type="scheduler"
+        )
```
As we've broken out into a helper function I would prefer to return these values rather than update the object.

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -393,6 +396,73 @@ def __init__(
+                        for k, v in self.env.items()
+                    ]
+                )
+            pod_template.metadata.generate_name = self._generate_name
```
Again I would prefer to return the template rather than overwrite values.

Timost replied:
As far as I can tell this requires to `copy.deepcopy` the template. Is that okay with you ?

jacobtomlinson replied:
Sure.

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -425,33 +495,8 @@ def __init__(
-                "docstring for ways to specify workers"
-            )
-            raise ValueError(msg)
+        self._clean_pod_templates()
```
This function is a bit **magic**. It's not obvious what the result of calling it is. I would prefer a direct assignment for clarity.

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -464,28 +509,14 @@ def __init__(
-                ]
-            )
-        self.pod_template.metadata.generate_name = self._generate_name
+        self._fill_pod_templates()
```
Again I see the need for breaking this out into a function but it's a bit magic.

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -393,6 +396,73 @@ def __init__(
         self.kwargs = kwargs
         super().__init__(**self.kwargs)
 
+    def _clean_pod_templates(self):
```
Can we have a docstring here?

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -393,6 +396,73 @@ def __init__(
+            self.scheduler_pod_template, pod_type="scheduler"
+        )
+
+    def _fill_pod_templates(self):
```
Can we have a docstring here?

Timost replied:
Maybe `_fill_pod_templates` and `_clean_pod_templates` (or there equivalent) should go in `objects.py` what do you think ?

jacobtomlinson replied:
I'm not sure if that would add any benefit. It just adds further indirection.

Timost replied:
Ok I'll leave them there then.

Timost commented:
@jacobtomlinson Could you give me a few pointers on how to run the tests on my computer ? looking at `.circleci/config.yml` it looks like I need to install minikube and run the tests against a k8s cluster inside minikube.

I'm using Debian.

jacobtomlinson commented:
Check out the testing section of the docs.

https://kubernetes.dask.org/en/latest/testing.html

Timost commented:
Oh thanks. I missed that sorry !

Timost commented:
PTAL:
- Changed methods style (return new values for assignement) thanks for the nice link.
- Added doc strings (not sure if this is enough)
- Fixed the code so the tests pass (at least on my computer)

jacobtomlinson commented:
That's great many thanks for making these changes!

Timost commented:
Thank you for the very responsive review ! It's really nice !
